web/api/v2/create-action/index.ts

```diff
@@ -0,0 +1,187 @@
+});
+
+export const POST = async (
+  req: NextRequest,
+  { params: rawParams }: { params: { app_id: string } },
+) => {
+  const rawBody = await req.json();
+  const api_key = req.headers.get("authorization")?.split(" ")[1];
+
+  if (!api_key) {
+    return errorResponse({
+      statusCode: 401,
+      code: "unauthorized",
+      detail: "API key is required.",
+      attribute: "api_key",
+      req,
+    });
+  }
+
+  const {
+    isValid: isParamsValid,
+    parsedParams: params,
+    handleError: handleParamsValidationError,
+  } = await validateRequestSchema({
+    schema: createActionParamsSchema,
+    value: rawParams,
+  });
+
+  if (!isParamsValid) {
+    return handleParamsValidationError(req);
+  }
+
+  const {
+    isValid: isBodyValid,
+    parsedParams: body,
+    handleError: handleBodyValidationError,
+  } = await validateRequestSchema({
+    schema: createActionBodySchema,
+    value: rawBody,
+  });
+
+  if (!isBodyValid) {
+    return handleBodyValidationError(req);
+  }
+
+  const { action, app_id } = {
+    ...body,
+    ...params,
+  };
+
+  const keyValue = api_key.replace(/^api_/, "");
+  const base64ApiKey = Buffer.from(keyValue, "base64").toString("utf8");
+  const [id, secret] = base64ApiKey.split(":");
+  const serviceClient = await getAPIServiceGraphqlClient();
+
+  const { api_key_by_pk } = await fetchApiKeySdk(
+    serviceClient,
+  ).VerifyFetchAPIKey({
+    id,
+    appId: app_id,
+  });
+
+  if (!api_key_by_pk) {
+    return errorResponse({
+      statusCode: 404,
+      code: "not_found",
+      detail: "API key not found.",
+      attribute: "api_key",
+      req,
+    });
+  }
+
+  if (!api_key_by_pk.is_active) {
+    return errorResponse({
+      statusCode: 400,
+      code: "api_key_inactive",
+      detail: "API key is inactive.",
+      attribute: "api_key",
+      req,
+    });
+  }
+
+  if (!api_key_by_pk.team.apps.some((a) => a.id === app_id)) {
+    return errorResponse({
+      statusCode: 403,
+      code: "invalid_app",
+      detail: "API key is not valid for this app.",
+      attribute: "api_key",
+      req,
+    });
+  }
+
+  const isAPIKeyValid = verifyHashedSecret(
+    api_key_by_pk.id,
+    secret,
+    api_key_by_pk.api_key,
+  );
+
+  if (!isAPIKeyValid) {
+    return errorResponse({
+      statusCode: 403,
+      code: "invalid_api_key",
+      detail: "API key is not valid.",
+      attribute: "api_key",
+      req,
+    });
+  }
+
+  const team_id = api_key_by_pk.team.id;
+  const apiKeyClient = await getAPIKeyGraphqlClient({ team_id });
+
+  const external_nullifier = generateExternalNullifier(
+    app_id as `app_${string}`,
+    action,
+  ).digest;
+
+  let insertedAction: CreateDynamicActionMutation["insert_action_one"] | null =
+    null;
+
+  try {
+    const { insert_action_one } = await createDynamicActionSdk(
+      apiKeyClient,
+    ).CreateDynamicAction({
+      app_id: app_id,
+      action: body.action,
+      external_nullifier,
```

Comment:
@andy-t-wang Should we also accept some additional params in this endpoint like `max_verification`?

Reply:
ah yeah that's a good idea